[clang-query] add basic profiling on matching each ASTs

danix800 opened the pull request:
We've found that basic profiling could help improving/optimizing when developing clang-tidy checks.

This PR adds an extra command
```
set enable-profile (true|false)   Set whether to enable matcher profiling.
```
which enables profiling queries on each file.

Sample output:

```
$ cat test.cql
set enable-profile true
m binaryOperator(isExpansionInMainFile())

$ cat test.c
int test(int i, int j) {
  return i + j;
}

$ clang-query --track-memory -f test.cql test.c --

Match #1:

{{.*}}/test.c:2:10: note: "root" binds here
    2 |   return i + j;
      |          ^~~~~
1 match.
===-------------------------------------------------------------------------===
                         clang-query matcher profiling
===-------------------------------------------------------------------------===
  Total Execution Time: 0.0000 seconds (0.0000 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  ---Mem---  --- Name ---
   0.0000 (100.0%)   0.0000 (100.0%)   0.0000 (100.0%)   0.0000 (100.0%)        224  {{.*}}/test.c
   0.0000 (100.0%)   0.0000 (100.0%)   0.0000 (100.0%)   0.0000 (100.0%)        224  Total
```

danix800 committed:
[clang-query] add basic profiling on matching each ASTs

Sample output:

$ cat test.cql
set enable-profile true
m binaryOperator(isExpansionInMainFile())

$ cat test.c
int test(int i, int j) {
  return i + j;
}

$ clang-query --track-memory -f test.cql test.c --

Match #1:

{{.*}}/test.c:2:10: note: "root" binds here
    2 |   return i + j;
      |          ^~~~~
1 match.
===-------------------------------------------------------------------------===
                         clang-query matcher profiling
===-------------------------------------------------------------------------===
  Total Execution Time: 0.0000 seconds (0.0000 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  ---Mem---  --- Name ---
   0.0000 (100.0%)   0.0000 (100.0%)   0.0000 (100.0%)   0.0000 (100.0%)        224  {{.*}}/test.c
   0.0000 (100.0%)   0.0000 (100.0%)   0.0000 (100.0%)   0.0000 (100.0%)        224  Total

danix800 committed:
fix typos

danix800 committed:
fix punctuation error in command help text

AaronBallman commented on clang-tools-extra/clang-query/QueryProfile.h:
```diff
@@ -0,0 +1,35 @@
+  ~QueryProfile();
+
+private:
+  std::optional<llvm::TimerGroup> TG;
```
Why is this `optional`?

danix800 replied:
No need to be optional, I'll remove it.

AaronBallman commented on clang-tools-extra/clang-query/QueryProfile.h:
```diff
@@ -0,0 +1,35 @@
+
+class QueryProfile {
+public:
+  llvm::StringMap<llvm::TimeRecord> Records;
```
Rather than leave this public, should we add an interface to insert a new record?

danix800 replied:
`QueryProfile` is a very small utility like class, I'll move it into Query.cpp.

danix800 replied:
QueryProfiler could be simplified to no more then 10 lines.

AaronBallman reviewed (commented):
This is really cool, thank you! Can you also add a release note to `clang-tools-extra/docs/ReleaseNotes.rst` so users know about the new functionality?

Changes generally look good, I just had a few questions.

danix800 committed:
move QueryProfiler into Query.cpp

1. QueryProfiler is very small, no need to expose it;
2. TimerGroup is created when dumping profiling info,
   no need to be an optional member of QueryProfiler.

danix800 committed:
QueryProfiler further simplifiled

danix800 committed:
update ReleaseNotes.rst for this new profiling feature

danix800 commented:
> This is really cool, thank you! Can you also add a release note to `clang-tools-extra/docs/ReleaseNotes.rst` so users know about the new functionality?
> 
> Changes generally look good, I just had a few questions.

ReleaseNotes updated.

AaronBallman reviewed (approved):
LGTM, thank you!
